Implemented mimic implementation of sntp from the original java implementation

hkh4n opened the pull request:
So basically all of this code is basically a direct copy translated from java to go from here: https://github.com/i2p/i2p.i2p/tree/master/router/java/src/net/i2p/router/time

Though this does not embed a whole sntp client in the code but instead uses https://github.com/beevik/ntp.

this sntp package implements the components necessary to implement router skew in the future.

RouterTimestamper:
- manages periodic synchronization with NTP servers.
- supports prioritization of NTP servers based on geographic zones.
- handles concurrent queries (just like the java implementation) to multiple NTP servers for improved accuracy.

Zones:
- self-explanatory, mapping for the appropriate NTP zones. embedded data for portability.

UpdateListener:
- SetNow mothod for listeners to receive updated time and stratums


Unit tests are included and of course pass on this build.

Apologies in advance for the single commit. There were so many commits that were done in combination of async work between two computers and constant edits and removing of unclean code that it would've looked pretty ugly on a PR so I've put it all in one commit :).

hkh4n committed:
!WIP! router_timestamper

eyedeekay commented:
Now that I've had some time to actually look at this I really like what you've done here, you've picked up on the structure we're trying to go for with the project outstandingly. I do however think we need to consider some changes and potentially simplifications, though, I shall attach some suggestions in the files.

eyedeekay commented on lib/util/time/sntp/router_timestamper.go:
```diff
@@ -0,0 +1,322 @@
+func (rt *RouterTimestamper) run() {
+	defer rt.waitGroup.Done()
+	lastFailed := false
+	for rt.isRunning {
```
I think we should factor out the section of code starting here into something like:

```Go
func (rt *RouterTimestamper) performTimeQuery() bool {
	rt.updateConfig()
	preferIPv6 := checkIPv6Connectivity()
	if rt.disabled {
		return false
	}
	
	if rt.priorityServers != nil {
		for _, servers := range rt.priorityServers {
			if !rt.queryTime(servers, shortTimeout, preferIPv6) {
				return true
			}
		}
	}
	
	prefIPv6 := preferIPv6 && rt.secureRandBool(0.75)
	return !rt.queryTime(rt.servers, defaultTimeout, prefIPv6)
}
```

Then use the return value to set `lastFailed` to de-duplicate this section of code, which appears again in `runOnce()`.

hkh4n replied:
Agreed, this was an oversight when producing the code and I hadn't realized it at the time. Refactored in https://github.com/go-i2p/go-i2p/pull/9/commits/f45d3018687782ddbaa5e50eca664a4054a4d615

eyedeekay commented on lib/util/time/sntp/router_timestamper.go:
```diff
@@ -0,0 +1,322 @@
+func (rt *RouterTimestamper) runOnce() {
+	lastFailed := false
+	rt.updateConfig()
+	preferIPv6 := checkIPv6Connectivity()
```
See above

eyedeekay commented on lib/util/time/sntp/router_timestamper.go:
```diff
@@ -0,0 +1,322 @@
+			fmt.Printf("NTP query failed: %v\n", err)
+			return false
+		}
+
```
Do we need to validate the response once we receive it with SNTP? Alternatively, since we might at some point need to have multiple ways of getting an accurate time, should we have one here for example when we design the interface? something like:

```Go
if (!rt.validateRespnonse(response)) {
    log.Error("Invalid SNTP response")
    return false

}
```

If so, what should go into a `validateResponse` function? I have some idea of the attacks which are possible on NTP, but when and how they translate to SNTP is something I'll have to catch up on.

hkh4n replied:
Hmm, I hadn't thought about attacks to be fair. The only thing in mind was to get a working implementation but I think we could verify through some sort of sanity check on timestamps and some sort of acceptable tolerance. An excessive delay would return false or of course something impossible like in the future. I can imagine creating threats of fuzz to see how resilient the sntp client really is.

edit: Got verification done here -> https://github.com/go-i2p/go-i2p/pull/9/commits/ade80e577cea0f3e89e073282cecad84e2294b1d

hkh4n committed:
proposed refactor
-added secureRandBool() and performTimeQuery()

hkh4n committed:
added sntp verification
-Leap Indicator
-Stratum level check
-Round-trip Delay
-Clock offset
-simple non-zero time
-Root Dispersion and Root Delay

hkh4n committed:
Merge branch 'master' into sntp-experimental

eyedeekay commented:
All right looks reasonable to me, merge time.
